Test case failures in Python 3.9

# Bug

`__args__` was removed in https://bugs.python.org/issue40397 and https://bugs.python.org/issue40398 . It seems a lot of places use it that causes test failures in Python 3.9. A quick grep of args shows many places. 

```
rg __args__ 
pydantic/fields.py
425:            for type_ in self.type_.__args__:
447:            for i, t in enumerate(self.type_.__args__):
449:                    self.type_ = self.type_.__args__[0]
463:            self.type_ = self.type_.__args__[0]
473:            self.type_ = self.type_.__args__[0]
476:            self.type_ = self.type_.__args__[0]
479:            self.type_ = self.type_.__args__[0]
482:            self.key_field = self._create_sub_type(self.type_.__args__[0], 'key_' + self.name, for_keys=True)
483:            self.type_ = self.type_.__args__[1]
488:            self.type_ = self.type_.__args__[0]
497:            self.sub_fields = [self._create_sub_type(t, f'{self.name}_{i}') for i, t in enumerate(self.type_.__args__)]

pydantic/types.py
135:    __args__: List[Type[T]]  # type: ignore
167:    # __args__ is needed to conform to typing generics api
168:    namespace = {'min_items': min_items, 'max_items': max_items, 'item_type': item_type, '__args__': [item_type]}
177:    __args__: Set[Type[T]]  # type: ignore
206:    # __args__ is needed to conform to typing generics api
207:    namespace = {'min_items': min_items, 'max_items': max_items, 'item_type': item_type, '__args__': [item_type]}

pydantic/schema.py
808:            args: Tuple[Any, ...] = type_.__args__

pydantic/typing.py
179:        return type_.__args__
250:            if type_.__args__ is None or not isinstance(type_.__args__[0], type):
252:            return type_.__args__[0]
```

**Sample test failure :** 

```
pytest -k test_list_bare  
=============================== test session starts ===============================
platform linux -- Python 3.9.0b5, pytest-5.3.1, py-1.5.3, pluggy-0.12.0
rootdir: /root/checked_repos/pydantic, inifile: setup.cfg, testpaths: tests
plugins: hypothesis-4.43.1, cov-2.8.1
collected 1754 items / 1753 deselected / 1 selected                               

tests/test_edge_cases.py F                                                  [100%]

==================================== FAILURES =====================================
_________________________________ test_list_bare __________________________________

    def test_list_bare():
>       class MyModel(BaseModel):

tests/test_edge_cases.py:1356: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pydantic/main.py:262: in __new__
    fields[ann_name] = inferred = ModelField.infer(
pydantic/fields.py:309: in infer
    return cls(
pydantic/fields.py:271: in __init__
    self.prepare()
pydantic/fields.py:345: in prepare
    self._type_analysis()
pydantic/fields.py:463: in _type_analysis
    self.type_ = self.type_.__args__[0]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = typing.List, attr = '__args__'

    def __getattr__(self, attr):
        # We are careful for copy and pickle.
        # Also for simplicity we just don't relay all dunder names
        if '__origin__' in self.__dict__ and not _is_dunder(attr):
            return getattr(self.__origin__, attr)
>       raise AttributeError(attr)
E       AttributeError: __args__

/usr/lib/python3.9/typing.py:640: AttributeError
======================= 1 failed, 1753 deselected in 1.37s ========================
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -15,7 +15,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: ['3.6', '3.7', '3.8']
+        python-version: ['3.6', '3.7', '3.8', '3.9']
     env:
       PYTHON: ${{ matrix.python-version }}
       OS: ubuntu

--- a/changes/1832-PrettyWood.md
+++ b/changes/1832-PrettyWood.md
@@ -1,0 +1,1 @@
+add basic support of python 3.9

--- a/changes/1832-PrettyWood.md
+++ b/changes/1832-PrettyWood.md
@@ -1,1 +1,1 @@
-add basic support of python 3.9
+add basic support for python 3.9

--- a/pydantic/fields.py
+++ b/pydantic/fields.py
@@ -32,6 +32,8 @@ from .typing import (
     NoArgAnyCallable,
     NoneType,
     display_as_type,
+    get_args,
+    get_origin,
     is_literal_type,
     is_new_type,
     new_type_supertype,
@@ -411,7 +413,7 @@ class ModelField(Representation):
         elif is_literal_type(self.type_):
             return
 
-        origin = getattr(self.type_, '__origin__', None)
+        origin = get_origin(self.type_)
         if origin is None:
             # field is not "typing" object eg. Union, Dict, List etc.
             # allow None for virtual superclasses of NoneType, e.g. Hashable
@@ -422,7 +424,7 @@ class ModelField(Representation):
             return
         if origin is Union:
             types_ = []
-            for type_ in self.type_.__args__:
+            for type_ in get_args(self.type_):
                 if type_ is NoneType:
                     if self.required is Undefined:
                         self.required = False
@@ -444,9 +446,9 @@ class ModelField(Representation):
         if issubclass(origin, Tuple):  # type: ignore
             self.shape = SHAPE_TUPLE
             self.sub_fields = []
-            for i, t in enumerate(self.type_.__args__):
+            for i, t in enumerate(get_args(self.type_)):
                 if t is Ellipsis:
-                    self.type_ = self.type_.__args__[0]
+                    self.type_ = get_args(self.type_)[0]
                     self.shape = SHAPE_TUPLE_ELLIPSIS
                     return
                 self.sub_fields.append(self._create_sub_type(t, f'{self.name}_{i}'))
@@ -460,7 +462,7 @@ class ModelField(Representation):
                     {f'list_{i}': Validator(validator, pre=True) for i, validator in enumerate(get_validators())}
                 )
 
-            self.type_ = self.type_.__args__[0]
+            self.type_ = get_args(self.type_)[0]
             self.shape = SHAPE_LIST
         elif issubclass(origin, Set):
             # Create self validators
@@ -470,22 +472,22 @@ class ModelField(Representation):
                     {f'set_{i}': Validator(validator, pre=True) for i, validator in enumerate(get_validators())}
                 )
 
-            self.type_ = self.type_.__args__[0]
+            self.type_ = get_args(self.type_)[0]
             self.shape = SHAPE_SET
         elif issubclass(origin, FrozenSet):
-            self.type_ = self.type_.__args__[0]
+            self.type_ = get_args(self.type_)[0]
             self.shape = SHAPE_FROZENSET
         elif issubclass(origin, Sequence):
-            self.type_ = self.type_.__args__[0]
+            self.type_ = get_args(self.type_)[0]
             self.shape = SHAPE_SEQUENCE
         elif issubclass(origin, Mapping):
-            self.key_field = self._create_sub_type(self.type_.__args__[0], 'key_' + self.name, for_keys=True)
-            self.type_ = self.type_.__args__[1]
+            self.key_field = self._create_sub_type(get_args(self.type_)[0], 'key_' + self.name, for_keys=True)
+            self.type_ = get_args(self.type_)[1]
             self.shape = SHAPE_MAPPING
         # Equality check as almost everything inherits form Iterable, including str
         # check for Iterable and CollectionsIterable, as it could receive one even when declared with the other
         elif origin in {Iterable, CollectionsIterable}:
-            self.type_ = self.type_.__args__[0]
+            self.type_ = get_args(self.type_)[0]
             self.shape = SHAPE_ITERABLE
             self.sub_fields = [self._create_sub_type(self.type_, f'{self.name}_type')]
         elif issubclass(origin, Type):  # type: ignore
@@ -494,7 +496,7 @@ class ModelField(Representation):
             # Is a Pydantic-compatible generic that handles itself
             # or we have arbitrary_types_allowed = True
             self.shape = SHAPE_GENERIC
-            self.sub_fields = [self._create_sub_type(t, f'{self.name}_{i}') for i, t in enumerate(self.type_.__args__)]
+            self.sub_fields = [self._create_sub_type(t, f'{self.name}_{i}') for i, t in enumerate(get_args(self.type_))]
             self.type_ = origin
             return
         else:

--- a/pydantic/generics.py
+++ b/pydantic/generics.py
@@ -3,6 +3,7 @@ from typing import TYPE_CHECKING, Any, ClassVar, Dict, Tuple, Type, TypeVar, Uni
 from .class_validators import gather_all_validators
 from .fields import FieldInfo, ModelField
 from .main import BaseModel, create_model
+from .typing import get_origin
 from .utils import lenient_issubclass
 
 _generic_types_cache: Dict[Tuple[Type[Any], Union[Any, Tuple[Any, ...]]], Type[BaseModel]] = {}
@@ -37,7 +38,7 @@ class GenericModel(BaseModel):
         check_parameters_count(cls, params)
         typevars_map: Dict[TypeVarType, Type[Any]] = dict(zip(cls.__parameters__, params))
         type_hints = get_type_hints(cls).items()
-        instance_type_hints = {k: v for k, v in type_hints if getattr(v, '__origin__', None) is not ClassVar}
+        instance_type_hints = {k: v for k, v in type_hints if get_origin(v) is not ClassVar}
         concrete_type_hints: Dict[str, Type[Any]] = {
             k: resolve_type_hint(v, typevars_map) for k, v in instance_type_hints.items()
         }
@@ -79,7 +80,7 @@ class GenericModel(BaseModel):
 
 
 def resolve_type_hint(type_: Any, typevars_map: Dict[Any, Any]) -> Type[Any]:
-    if hasattr(type_, '__origin__') and getattr(type_, '__parameters__', None):
+    if get_origin(type_) and getattr(type_, '__parameters__', None):
         concrete_type_args = tuple([typevars_map[x] for x in type_.__parameters__])
         return type_[concrete_type_args]
     return typevars_map.get(type_, type_)

--- a/pydantic/main.py
+++ b/pydantic/main.py
@@ -33,7 +33,7 @@ from .json import custom_pydantic_encoder, pydantic_encoder
 from .parse import Protocol, load_file, load_str_bytes
 from .schema import model_schema
 from .types import PyObject, StrBytes
-from .typing import AnyCallable, ForwardRef, is_classvar, resolve_annotations, update_field_forward_refs
+from .typing import AnyCallable, ForwardRef, get_origin, is_classvar, resolve_annotations, update_field_forward_refs
 from .utils import (
     ClassAttribute,
     GetterDict,
@@ -256,7 +256,7 @@ class ModelMetaclass(ABCMeta):
                     if (
                         isinstance(value, untouched_types)
                         and ann_type != PyObject
-                        and not lenient_issubclass(getattr(ann_type, '__origin__', None), Type)
+                        and not lenient_issubclass(get_origin(ann_type), Type)
                     ):
                         continue
                     fields[ann_name] = inferred = ModelField.infer(

--- a/pydantic/schema.py
+++ b/pydantic/schema.py
@@ -55,7 +55,7 @@ from .types import (
     conset,
     constr,
 )
-from .typing import ForwardRef, Literal, is_callable_type, is_literal_type, literal_values
+from .typing import ForwardRef, Literal, get_args, get_origin, is_callable_type, is_literal_type, literal_values
 from .utils import get_model, lenient_issubclass, sequence_like
 
 if TYPE_CHECKING:
@@ -803,9 +803,9 @@ def get_annotation_from_field_info(annotation: Any, field_info: FieldInfo, field
             or lenient_issubclass(type_, (ConstrainedList, ConstrainedSet))
         ):
             return type_
-        origin = getattr(type_, '__origin__', None)
+        origin = get_origin(type_)
         if origin is not None:
-            args: Tuple[Any, ...] = type_.__args__
+            args: Tuple[Any, ...] = get_args(type_)
             if any(isinstance(a, ForwardRef) for a in args):
                 # forward refs cause infinite recursion below
                 return type_

--- a/pydantic/typing.py
+++ b/pydantic/typing.py
@@ -17,6 +17,7 @@ from typing import (  # type: ignore
     Type,
     Union,
     _eval_type,
+    cast,
 )
 
 try:
@@ -53,7 +54,9 @@ elif sys.version_info < (3, 9):
 else:
 
     def evaluate_forwardref(type_: ForwardRef, globalns: Any, localns: Any) -> Any:
-        return type_._evaluate(globalns, localns, set())
+        # Even though it is the right signature for python 3.9, mypy complains with
+        # `error: Too many arguments for "_evaluate" of "ForwardRef"` hence the cast...
+        return cast(Any, type_)._evaluate(globalns, localns, set())
 
 
 if sys.version_info < (3, 7):

--- a/pydantic/typing.py
+++ b/pydantic/typing.py
@@ -84,6 +84,17 @@ else:
     def get_origin(tp: Type[Any]) -> Type[Any]:
         return typing_get_origin(tp) or getattr(tp, '__origin__', None)
 
+    def generic_get_args(tp: Type[Any]) -> Tuple[Any, ...]:
+        """
+        In python 3.9, `typing.Dict`, `typing.List`, ...
+        do have an empty `__args__` by default (instead of the generic ~T for example).
+        In order to still support `Dict` for example and consider it as `Dict[Any, Any]`,
+        we retrieve the `_nparams` value that tells us how many parameters it needs.
+        """
+        if hasattr(tp, '_nparams'):
+            return (Any,) * tp._nparams
+        return ()
+
     def get_args(tp: Type[Any]) -> Tuple[Any, ...]:
         """Get type arguments with all substitutions performed.
 
@@ -99,7 +110,7 @@ else:
             args = typing_get_args(tp)
         except IndexError:
             args = ()
-        return args or getattr(tp, '__args__', ())
+        return args or getattr(tp, '__args__', ()) or generic_get_args(tp)
 
 
 if TYPE_CHECKING:

--- a/pydantic/typing.py
+++ b/pydantic/typing.py
@@ -89,6 +89,12 @@ else:
     from typing import Literal, get_args as typing_get_args, get_origin as typing_get_origin
 
     def get_origin(tp: Type[Any]) -> Type[Any]:
+        """
+        We can't directly use `typing.get_origin` since we need a fallback to support
+        custom generic classes like `ConstrainedList`
+        It should be useless once https://github.com/cython/cython/issues/3537 is
+        solved and https://github.com/samuelcolvin/pydantic/pull/1753 is merged.
+        """
         return typing_get_origin(tp) or getattr(tp, '__origin__', None)
 
     def generic_get_args(tp: Type[Any]) -> Tuple[Any, ...]:
@@ -119,6 +125,7 @@ else:
         # TODO: remove the pragma: no cover once we can run CI on python 3.9
         except IndexError:  # pragma: no cover
             args = ()
+        # the fallback is needed for the same reasons as `get_origin` (see above)
         return args or getattr(tp, '__args__', ()) or generic_get_args(tp)
 
 

--- a/pydantic/typing.py
+++ b/pydantic/typing.py
@@ -44,10 +44,16 @@ if sys.version_info < (3, 7):
         return type_._eval_type(globalns, localns)
 
 
-else:
+elif sys.version_info < (3, 9):
 
     def evaluate_forwardref(type_: ForwardRef, globalns: Any, localns: Any) -> Any:
         return type_._evaluate(globalns, localns)
+
+
+else:
+
+    def evaluate_forwardref(type_: ForwardRef, globalns: Any, localns: Any) -> Any:
+        return type_._evaluate(globalns, localns, set())
 
 
 if sys.version_info < (3, 7):

--- a/pydantic/typing.py
+++ b/pydantic/typing.py
@@ -52,7 +52,8 @@ elif sys.version_info < (3, 9):
 
 else:
 
-    def evaluate_forwardref(type_: ForwardRef, globalns: Any, localns: Any) -> Any:
+    # TODO: remove the pragma: no cover once we can run CI on python 3.9
+    def evaluate_forwardref(type_: ForwardRef, globalns: Any, localns: Any) -> Any:  # pragma: no cover
         return type_._evaluate(globalns, localns, set())
 
 
@@ -97,7 +98,8 @@ else:
         In order to still support `Dict` for example and consider it as `Dict[Any, Any]`,
         we retrieve the `_nparams` value that tells us how many parameters it needs.
         """
-        if hasattr(tp, '_nparams'):
+        # TODO: remove the pragma: no cover once we can run CI on python 3.9
+        if hasattr(tp, '_nparams'):  # pragma: no cover
             return (Any,) * tp._nparams
         return ()
 
@@ -114,7 +116,8 @@ else:
         """
         try:
             args = typing_get_args(tp)
-        except IndexError:
+        # TODO: remove the pragma: no cover once we can run CI on python 3.9
+        except IndexError:  # pragma: no cover
             args = ()
         return args or getattr(tp, '__args__', ()) or generic_get_args(tp)
 

--- a/pydantic/typing.py
+++ b/pydantic/typing.py
@@ -70,8 +70,37 @@ if sys.version_info < (3, 8):
             from typing_extensions import Literal
         except ImportError:
             Literal = None
+
+    def get_args(t: Type[Any]) -> Tuple[Any, ...]:
+        return getattr(t, '__args__', ())
+
+    def get_origin(t: Type[Any]) -> Optional[Type[Any]]:
+        return getattr(t, '__origin__', None)
+
+
 else:
-    from typing import Literal
+    from typing import Literal, get_args as typing_get_args, get_origin as typing_get_origin
+
+    def get_origin(tp: Type[Any]) -> Type[Any]:
+        return typing_get_origin(tp) or getattr(tp, '__origin__', None)
+
+    def get_args(tp: Type[Any]) -> Tuple[Any, ...]:
+        """Get type arguments with all substitutions performed.
+
+        For unions, basic simplifications used by Union constructor are performed.
+        Examples::
+            get_args(Dict[str, int]) == (str, int)
+            get_args(int) == ()
+            get_args(Union[int, Union[T, int], str][int]) == (int, str)
+            get_args(Union[int, Tuple[T, int]][str]) == (int, Tuple[str, int])
+            get_args(Callable[[], T][int]) == ([], int)
+        """
+        try:
+            args = typing_get_args(tp)
+        except IndexError:
+            args = ()
+        return args or getattr(tp, '__args__', ())
+
 
 if TYPE_CHECKING:
     from .fields import ModelField
@@ -115,6 +144,8 @@ __all__ = (
     'CallableGenerator',
     'ReprArgs',
     'CallableGenerator',
+    'get_args',
+    'get_origin',
 )
 
 
@@ -167,16 +198,16 @@ def resolve_annotations(raw_annotations: Dict[str, Type[Any]], module_name: Opti
 
 
 def is_callable_type(type_: Type[Any]) -> bool:
-    return type_ is Callable or getattr(type_, '__origin__', None) is Callable
+    return type_ is Callable or get_origin(type_) is Callable
 
 
 if sys.version_info >= (3, 7):
 
     def is_literal_type(type_: Type[Any]) -> bool:
-        return Literal is not None and getattr(type_, '__origin__', None) is Literal
+        return Literal is not None and get_origin(type_) is Literal
 
     def literal_values(type_: Type[Any]) -> Tuple[Any, ...]:
-        return type_.__args__
+        return get_args(type_)
 
 
 else:
@@ -217,12 +248,15 @@ def new_type_supertype(type_: Type[Any]) -> Type[Any]:
     return type_
 
 
-def _check_classvar(v: Type[Any]) -> bool:
+def _check_classvar(v: Optional[Type[Any]]) -> bool:
+    if v is None:
+        return False
+
     return v.__class__ == ClassVar.__class__ and (sys.version_info < (3, 7) or getattr(v, '_name', None) == 'ClassVar')
 
 
 def is_classvar(ann_type: Type[Any]) -> bool:
-    return _check_classvar(ann_type) or _check_classvar(getattr(ann_type, '__origin__', None))
+    return _check_classvar(ann_type) or _check_classvar(get_origin(ann_type))
 
 
 def update_field_forward_refs(field: 'ModelField', globalns: Any, localns: Any) -> None:
@@ -243,13 +277,13 @@ def get_class(type_: Type[Any]) -> Union[None, bool, Type[Any]]:
     without brackets. Otherwise returns None.
     """
     try:
-        origin = getattr(type_, '__origin__')
+        origin = get_origin(type_)
         if origin is None:  # Python 3.6
             origin = type_
         if issubclass(origin, Type):  # type: ignore
-            if type_.__args__ is None or not isinstance(type_.__args__[0], type):
+            if not get_args(type_) or not isinstance(get_args(type_)[0], type):
                 return True
-            return type_.__args__[0]
-    except AttributeError:
+            return get_args(type_)[0]
+    except (AttributeError, TypeError):
         pass
     return None

--- a/pydantic/typing.py
+++ b/pydantic/typing.py
@@ -52,8 +52,7 @@ elif sys.version_info < (3, 9):
 
 else:
 
-    # TODO: remove the pragma: no cover once we can run CI on python 3.9
-    def evaluate_forwardref(type_: ForwardRef, globalns: Any, localns: Any) -> Any:  # pragma: no cover
+    def evaluate_forwardref(type_: ForwardRef, globalns: Any, localns: Any) -> Any:
         return type_._evaluate(globalns, localns, set())
 
 
@@ -104,8 +103,7 @@ else:
         In order to still support `Dict` for example and consider it as `Dict[Any, Any]`,
         we retrieve the `_nparams` value that tells us how many parameters it needs.
         """
-        # TODO: remove the pragma: no cover once we can run CI on python 3.9
-        if hasattr(tp, '_nparams'):  # pragma: no cover
+        if hasattr(tp, '_nparams'):
             return (Any,) * tp._nparams
         return ()
 
@@ -122,8 +120,7 @@ else:
         """
         try:
             args = typing_get_args(tp)
-        # TODO: remove the pragma: no cover once we can run CI on python 3.9
-        except IndexError:  # pragma: no cover
+        except IndexError:
             args = ()
         # the fallback is needed for the same reasons as `get_origin` (see above)
         return args or getattr(tp, '__args__', ()) or generic_get_args(tp)

--- a/setup.cfg
+++ b/setup.cfg
@@ -17,7 +17,7 @@ per-file-ignores =
     docs/examples/types_constrained.py: F722
 
 [bdist_wheel]
-python-tag = py36.py37.py38
+python-tag = py36.py37.py38.py39
 
 [coverage:run]
 source = pydantic

--- a/setup.py
+++ b/setup.py
@@ -105,6 +105,7 @@ setup(
         'Programming Language :: Python :: 3.6',
         'Programming Language :: Python :: 3.7',
         'Programming Language :: Python :: 3.8',
+        'Programming Language :: Python :: 3.9',
         'Intended Audience :: Developers',
         'Intended Audience :: Information Technology',
         'Intended Audience :: System Administrators',

